### ISSUE TITLE
Unrecognized Device

### ISSUE THREAD

**Issue by stevenyeunghk** [superhawk610/multicast#44]
Hi all,

I have configured channel and registered chromecast in multicast, but the TV output unrecognized device name and the device id is n/a. My chromecast device has been already registered as developer device. Here is my environment:

Ubuntu 16.04 with firewall disabled (same network with chromecast device)
Node 6.14.2
MongoDB 3.6.5

Appreciate your help, many thanks

**Comment by superhawk610** [comment]
Are you able to add a Channel and set the Chromecast to that channel from the web interface?

Does the unrecognized device name / device id persist through restarting multicast?

Can you dump the contents of your Mongo DB using the following commands and reply with them? (Feel free to censor any personal information from the results):

```shell
nano dump.js

### dump.js #####
use multicast
db.chromecasts.find()
#################

mongo -u <USERNAME> -p <PASSWORD> --authenticationDatabase admin < dump.js
```

**Comment by stevenyeunghk** [comment]
Thanks for your reply. I'm able to add a channel and assign the chromecast with that channel from multicast web interface. 

Both collections "channels" and "chromecasts" in MongoDB with all values set same as the values set in web interface. 

However, the TV just shows a light green background with the message "Welcome to Multicast...Device is unregistered... the device id is n/a and the device name is unrecognized". No matter I restart the multicast, the TV still can't show the page from URL defined in the channel.

**Comment by superhawk610** [comment]
Hmm. Since I'm unable to replicate your issue, you'll have to debug on your end. The first thing I would try is to make the following edits:

```diff
// app/main.js, line 59

/* Landing Page */
app.get('/landing', (req, res) => {
  var ip = stripIPv6(req.connection.remoteAddress), // Get IPv4 address of device
    d = devices.withHost(ip) // Find local info for device

+ // DEBUG ==========
+ const now = Date.now()
+ console.log(now, 'remoteAddress', req.connection.remoteAddress)
+ console.log(now, 'ip', ip)
+ console.log(now, 'device', d)
+ // END DEBUG ======

  if (d)
    res.redirect(`/devices/${d.deviceId}`) // Redirect to device display // Display device info
  else
    res.render('setup-chromecast', {
      device: {
        address: ip,
        deviceId: 'n/a',
        name: 'Unrecognized Device'
      },
      registered: false,
      setupUrl: `${req.protocol}://${req.hostname}:${port}/`
    })
})
```

then restart multicast, attempt to connect, and post your log results.

You're seeing that green screen because multicast either doesn't recognize your device's IP address or the IP address is being malformed somehow.

**Comment by blackjid** [comment]
hi @superhawk610 

I'm getting a similar problem, that's why I didn't want to start a new issue.
I'm running multicast on my mac, and I have a Google Wifi network.

If I add the debug lines you suggested, I get the following when restarting multicast
```
MultiCast is live!
Listening at port 3944...
1531517951117 'remoteAddress' '::ffff:192.168.86.1'
1531517951117 'ip' '192.168.86.1'
1531517951117 'device' undefined
connected client with host: 192.168.86.1
```

Also when I register my chromecast for the first time, the device on the chromecast page is never connected (green dot) and nothing changes on the chromecast screen. Only after I restart multicast, the green screen is displayed in the chromecast as the original poster described, without device id, no name and the gateway ip as the Network location

Any idea what else to do?

**Comment by superhawk610** [comment]
Let me kick around some ideas over the weekend and get back to you. I want to write a simple test suite to make sure it's actually a connection error and not something nasty hiding in Multicast I'm unaware of. I'll let you know once I have some tests for you to run to help diagnose this.

**Comment by blackjid** [comment]
I've been testing a bit, and I think the problem is in this function

https://github.com/superhawk610/multicast/blob/master/app/main.js#L65

It is not returning the device.

In a quick hardcoded test, it works fine if I change `d =  devices.withHost(ip)` with `d = devices.list()[0];` (I know that is the device I'm trying to get, the only one registered)

**Comment by blackjid** [comment]
Also a had to change this line https://github.com/superhawk610/multicast/blob/master/app/main.js#L64 to use `localAddress` instead of `remoteAddress` with local address I get the chromecast local IP, with remote address I get the gateway (192.168.86.1)

**Comment by superhawk610** [comment]
When using `localAddress` instead of `remoteAddress`, do you have full functionality? Or do you still have to hardcode the device index?

**Comment by blackjid** [comment]
I still have to hard code the index. The `withHost` method return undefined with `remoteAddress` or `localAddress`.

Looks like the `devices` is not yet populated With the devices. Like a timing issue, but not sure.

**Comment by blackjid** [comment]
Do you have an idea where I can look to try to get the correct devices??

**Comment by superhawk610** [comment]
@blackjid I'm moving apartments tonight but I'll try to write up the flow for the app later this evening.

**Comment by blackjid** [comment]
Hey! Awesome thanks..

**Comment by blackjid** [comment]
I don't know if this has anything to do or not.. but..

The `update` function in https://github.com/superhawk610/multicast/blob/master/app/lib/devices.js#L35-L38 seems to try to add some data to the registered devices when is called here https://github.com/superhawk610/multicast/blob/master/app/lib/devices.js#L115-L118. 

If I call `devices.list()` I get a list with registered and non registered devices. The unregistered devices have their address, but the registered ones don't. It made me think that's why the `devices.withHost` function is not returning anything from when called in the landing route.

All is pure speculation, nothing really tested...

**Comment by superhawk610** [comment]
Ok, so here's the general `devices` data flow from application startup:

```js
// main.js, line 32
dbConnect(config)
/**
 * Establish a connection with the Mongo instance and get a list of saved devices.
 * Looking in models/Chromecast.js, you see that each stored device contains the
 * following info:
 * {
 *   deviceId: string, // unique hex identifier provided by device itself
 *   location: string, // Location info for device (Living Room, etc) entered during registration
 *   rotation: string, // Screen orientation for device
 *   channel: Channel // reference to active Channel for device
 * }
 */

// lib/devices.js, lines 95 - 147
const findDevices = () => { ... }
/**
 * Browse local network for active devices. When a device is found ( browser.on('serviceUp') ),
 * examine its hex identifier ( service.name.split('-').pop ). If that identifier matches any of the
 * saved devices, append that device's network information to its entry in devices.list(). If not,
 * add it as an unregistered device.
 *
 * Devices may be easily identified to the user by their Chromecast name selected when initially
 * setting up the Chromecast for function (not though Multicast). This information is stored in
 * service.txtRecord.fn.
 *
 * A device's network location (IP address) can be determined from service.addresses[0] when it is
 * first discovered. This is used to tie the device's identifier to the socket.io channel established for
 * communication, as the only information provided by a socket when it connects is its network location.
```

After further inspection, it seems like the issue may be occurring during this `findDevices()` function. To diagnose this, add log statements before and after it is called during `devices.init()`:

```diff
// lib/devices.js, line 20
+ console.log(func.list())
findDevices()
+ console.log(func.list())
```

If `devices.list()` contains duplicate entries and doesn't join corresponding registered/unregistered devices together into the same entry, it's possible that there is something going wrong when comparing their hex IDs. That is done here:

```js
// lib/devices.js, line 29
isRegistered: id => devices.findIndex(d => d.deviceId == id) > -1,
```

You can extend this to add some debug logging and see if this is where your problem lies by doing something like the following:

```js
// lib/devices.js, line 29
isRegistered: id => {
  console.log(`checking registration for ${id} (${typeof id})`)
  const index = devices.findIndex(d => d.deviceId == id)
  console.log(`devices returned index ${index}`)
  return index > -1
}
```

The `findIndex` function searches through an array and returns the index of the first element that returns truthy for the given test condition (in this case, whether the device's `deviceId` matches the given `id`. It's possible that those IDs are different in some unexpected way? It may have something to do with the id generation here:

```js
// lib/devices.js, line 111
while (_id.length > 0 && id.length < 3) id = _id.pop()
```

I hope this at least gave you some new ideas to try!

**Comment by blackjid** [comment]
Wow, thanks @superhawk610 ..

It's super embarrassing, but now is working fine.... 
I wasn't using my LOCAL_IP when setting my cast application. I didn't read well enough. Without knowing how the cast flow works I thought that google needed to be able to reach the multicast server so I went ahead I used my public ip, port forwarding, and all that jazz. Also tried with a proxy :(

Thanks for your time and the awesome project...
